### ISSUE TITLE
Rule blocked: CORERULES-9632 - operation to extract distinct set of substrings

### ISSUE THREAD

**Issue by ASL-rmarshall** [cdisc-org/cdisc-rules-engine#1100]
**Links to related JIRA Tickets**
- https://jira.cdisc.org/projects/CORERULES/issues/CORERULES-9632

**Rule Information**
- **Standard**: USDM
- **Rule ID**: DDF00089
- **Rule Description**: Any parameter name referenced in a tag in the text should be specified in the data dictionary parameter maps.

**Describe the problem**
Some USDM classes contain an HTML-formatted attribute that may contain one or more "usdm:tag" references to parameter names.  These references will be in the form `<usdm:tag name="<ParameterName>"/>` or `<usdm:tag name="<ParameterName>"></usdm:tag>`. For example, the following text value contains two tags:
> Subjects shall be between <usdm:tag name="min_age"/> and <usdm:tag name="max_age"/>

This rule is to check that the value of the `name` attribute of every "usdm:tag" in the text matches a value of the `tag` attribute in an of the parameter maps associated with the referenced dictionary.  For the above example, the record should be reported if either "min_age" or "max_age" is not a value of the `tag` attribute for any record in `ParameterMap` dataset where the `parent_id` value matches the value of the `dictionaryId` attribute on the record being checked.

**Describe the solution**
To do this, I think there needs to be an operation that can extract a unique list of values for the `name` attribute of every `usdm:tag` element referenced in a specified variable (optionally within records grouped by specified variable(s)).  This could be done by HTML/XHTML parsing, regex, or something else. This list could then by compared (e.g., using something like the `contains_all` operator) with a distinct list of `tag` values that are grouped by dictionary id.

The proposed rule logic below shows two hypothetical operations:
- `distinct_string_part`, which uses a regular expression to extract a distinct list of substrings and has the following parameters (in addition to the standard `domain`, `name`, `id`, `group`, `group_aliases` parameters):
    - `regex`: a regular expression to specify the substring to extract
    - `match_group`: (optional, default=1) to specify which set of parentheses in the `regex` value contains the substring to extract. This is needed because additional sets of parentheses might be needed for pattern matching - for example, the second set of parentheses in the example below is used to specify an either/or match.
    - `filter`: as requested for the `distinct` operation.
- `distinct_html_part`, which uses an XPath expression to extract a distinct list of substrings from HTML text and has the following parameters (in addition to the standard `domain`, `name`, `id`, `group`, `group_aliases` parameters):
    - `xpath`: an XPath expression to locate the substrings. Note that the "usdm" namespace will probably not be specified in each piece of text so this might need to be added if needed by the parser.
 
Only one of these operations would be needed.

**Proposed rule logic**
```yaml
Check:
  any:
    - name: $valid_parameter_tag_names
      operator: not_contains_all
      value: $tag_names_by_regex
    - name: $valid_parameter_tag_names
      operator: not_contains_all
      value: $tag_names_by_html
Operations:
  - domain: ParameterMap
    group:
      - parent_id
    group_aliases:
      - dicitonaryId
    id: $valid_parameter_tag_names
    name: tag
    operator: distinct
  - group:
      - id
    id: $tag_names_by_regex
    match_group: 1
    name: text
    operator: distinct_string_part
    regex: '<usdm:tag name=\"([^\"]+)\"(\/>|><\/usdm:tag>)'
  - group:
      - id
    id: $tag_names_by_html
    name: text
    operator: distinct_html_part
    xpath: '//usdm:tag/@name'
```

**Comment by ASL-rmarshall** [comment]
JSONata can be used to extract multiple substrings using regex and check them, so this new operation is no longer needed.
